src/v/cluster/tests/id_allocator_stm_test.cc

```diff
@@ -22,112 +23,160 @@
 #include "test_utils/async.h"
 #include "test_utils/fixture.h"
 #include "test_utils/scoped_config.h"
+#include "tests/stm_test_fixture.h"
 
 #include <seastar/core/abort_source.hh>
 #include <seastar/core/future.hh>
 #include <seastar/core/sleep.hh>
 #include <seastar/util/defer.hh>
 #include <seastar/util/log.hh>
 
-#include <boost/range/irange.hpp>
-#include <boost/test/tools/old/interface.hpp>
-
 #include <thread>
 
 using namespace std::chrono_literals;
 
+namespace cluster {
+namespace {
+
 ss::logger idstmlog{"idstm-test"};
 
-struct id_allocator_stm_fixture : simple_raft_fixture {
-    void create_stm_and_start_raft() {
+struct id_allocator_stm_fixture : state_machine_fixture {
+    ss::future<> initialize_state_machines() {
         // set configuration parameters
         test_local_cfg.get("id_allocator_batch_size").set_value(int16_t(1));
         test_local_cfg.get("id_allocator_log_capacity").set_value(int16_t(2));
-        create_raft();
-        raft::state_machine_manager_builder stm_m_builder;
 
-        _stm = stm_m_builder.create_stm<cluster::id_allocator_stm>(
-          idstmlog, _raft.get(), config::shard_local_cfg());
+        create_nodes();
+        co_await start_nodes();
+    }
+
+    ss::future<> start_node(raft_node_instance& node) {
+        co_await node.initialise(all_vnodes());
+        raft::state_machine_manager_builder builder;
+        auto stm = builder.create_stm<cluster::id_allocator_stm>(
+          idstmlog, node.raft().get(), config::shard_local_cfg());
+        co_await node.start(std::move(builder));
+        node_stms.emplace(node.get_vnode(), std::move(stm));
+    }
+
+    ss::future<> start_nodes() {
+        co_await parallel_for_each_node(
+          [this](raft_node_instance& node) { return start_node(node); });
+    }
+
+    ss::future<> stop_and_recreate_nodes() {
+        absl::flat_hash_map<model::node_id, ss::sstring> data_directories;
+        for (auto& [id, node] : nodes()) {
+            data_directories[id]
+              = node->raft()->log()->config().base_directory();
+            node_stms.erase(node->get_vnode());
+        }
+
+        co_await ss::parallel_for_each(
+          std::views::keys(data_directories),
+          [this](model::node_id id) { return stop_node(id); });
+
+        for (auto& [id, data_dir] : data_directories) {
+            add_node(id, model::revision_id(0), std::move(data_dir));
+        }
+    }
+
+    ss::future<> reset(int64_t id) {
+        auto result = co_await retry_with_leader(
+          model::timeout_clock::now() + 30s,
+          [this, id](raft_node_instance& leader_node) {
+              auto stm = node_stms[leader_node.get_vnode()];
+              return stm->reset_next_id(id, 1s);
+          });
+        vassert(result.has_value(), "raft error");
```

Comment:
nit: use a test friendly macro? ASSERT_EQ_CORO or some such. (below too)

Reply:
vassert() crashes the process, so no further tests in this binary will run, thats the issue I was thinking about...

> The "below too" i

oic, I didn't notice the return type of the function when i made the comment. Per [gtest docs](https://google.github.io/googletest/advanced.html#assertion-placement) one way to avoid this is by changing the helper function signatures to return the result via method parameters (pass a reference etc)... I feel crashing the entire test at first failure is weird. feel free to skip if its too much work.